misc-tools/import-contest

```diff
@@ -0,0 +1,214 @@
+    return answer == "y"
+
+
+def usage():
+    print(f"Usage '{sys.argv[0]} <domjudge-api-url>")
+    exit(1)
+
+def parse_api_response(name: str, response: requests.Response):
+    # The connection worked, but we may have received an HTTP error
+    if response.status_code >= 300:
+        print(response.text)
+        if response.status_code == 401:
+            raise RuntimeError(
+                "Authentication failed, please check your DOMjudge credentials in ~/.netrc.")
+        else:
+            raise RuntimeError(
+                f"API request {name} failed (code {response.status_code}).")
+
+    # We got a successful HTTP response. It worked. Return the full response
+    return json.loads(response.text)
+
+def do_api_request(name: str):
+    '''Perform an API call to the given endpoint and return its data.
+
+    Parameters:
+        name (str): the endpoint to call
+
+    Returns:
+        The endpoint contents.
+
+    Raises:
+        RuntimeError when the response is not JSON or the HTTP status code is non 2xx.
+    '''
+
+    url = f'{api_url}/{name}'
+
+    try:
+        response = requests.get(url, headers=headers)
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError(e)
+
+    return parse_api_response(name, response)
+
+
+def upload_file(name: str, apifilename: str, file: str, data: dict = {}):
+    '''Upload the given file to the API at the given path with the given name.
+
+    Parameters:
+        name (str): the endpoint to call
+        apifilename (str): the argument name for the file to upload
+        file (str): the file to upload
+
+    Returns:
+        The parsed endpoint contents.
+
+    Raises:
+        RuntimeError when the HTTP status code is non 2xx.
+    '''
+
+    files = [(apifilename, open(file, "rb"))]
+
+    url = f"{api_url}/{name}"
+
+    response = requests.post(url, files=files, headers=headers, data=data)
+
+    return parse_api_response(name, response)
+
+
+def import_file(basename: str, files: List[str]):
+    any_matched = False
+    for file in files:
+        if os.path.exists(file):
+            if confirm(f"Import {basename}s from {file}?", False):
+                print(f"Importing {basename}s.")
+                type = file.split(".")[1]
+                response = upload_file(f"/users/{basename}s", type, file)
+                print(json.dumps(response, indent=4))
+            else:
+                print(f"Skipping {basename}s import.")
+            any_matched = True
+    if not any_matched:
+        if len(files) == 1:
+            print(f"'{files[0]}' not found, skipping {basename}s import.")
+        else:
+            prefix = [f"'{item}'" for item in files[:-1]]
+            prefix_joined = ", ".join(prefix)
+            last = f"'{files[-1]}'"
+            print(
+                f"Neither {prefix_joined} nor {last} found, skipping {basename}s import.")
+
+
+if len(sys.argv) < 2:
+    usage()
+
+api_url = sys.argv[1]
+
```

Comment:
As a first step here, we might want to check whether we have the right credentials and permissions by calling `{api_url}/user`?

Reply:
We didn't before but I do like it. Better to fail early. Will add.